plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/KGraphIncrementalDiagramGenerator.xtend

```diff
@@ -0,0 +1,753 @@
+        val maxLevel = hierarchyDepth // controls how many hierarchy levels should be generated
+        var currentLevel = 0
+        while (childrenToProcess.peek() !== null && currentLevel < maxLevel) {
+            
+            var elementsOnLevel = childrenToProcess.size()
+            while (elementsOnLevel > 0) {
+                processNextElement()
+            
+                elementsOnLevel--
+            }
+            currentLevel++
+            
+        }
+        * 
+        */
+
+        return if (cancelIndicator.canceled) 
+               null
+           else 
+               diagramRoot
+    }
+    
+    // TODO: need mechanisms to manipulate the queue in such a way to prioritize generating requested pieces
+    /**
+     * Function to request next part of the diagram. Should eventually support getting specific pieces if they are ready.
+     */
+    def SModelElement getNextDiagramPiece() {
+        if (childrenToProcess.peek() !== null) {
+            return processNextElement()
+        }
+    }
+    
+    /**
+     * Like getNextDiagramPiece, but doesn't return piece. Only triggers generation.
+     */
+    def generateNextDiagramPiece() {
+        getNextDiagramPiece
+    }
+    
+    def boolean nodeChildrenAllProcessed(KGraphElement node) {
+        for (child: childrenToProcess) {
+            if (child.parent.equals(node)) {
+                return false
+            }
+        }
+        return true
+    }
+    
+    private def SModelElement processNextElement() {
+        val node = childrenToProcess.remove()
+        // get parent node to add children to
+        val skNode = kGraphToSModelElementMap.get(node.parent)
+        
+        skNode.children.addAll(incrementalCreateNodesAndPrepareEdges(node, skNode))
+        incrementalPostProcess()
+        return skNode
+    }
+    
+    /**
+     * Translates one {@code node} and its outgoing edges to {@link SModelElement}s. Also handles tracing and
+     * mapping between {@link KGraphElement}s and SModelElements.
+     * The edges are translated together with the nodes, because {@link KNode}s contain {@link KEdge}s in the field 
+     * {@link KNode#getOutgoingEdges} as children, whereas outgoing {@link SEdge}s are siblings of their originating 
+     * {@link SNode}s.
+     * The edges are stored to be generated later in the {@code edgesToGenerate} list and need to be generated later.
+     * The children of the node are added to the {@code childrenToProcess} queue to be generated later.
+     */
+    private def List<SModelElement> incrementalCreateNodesAndPrepareEdges(KNode node, SModelElement parent) {
+        val nodeAndEdgeElements = new ArrayList
+        // add all node children
+        val SNode nodeElement = incrementalGenerateNode(node)
+        nodeAndEdgeElements.add(nodeElement)
+        kGraphToSModelElementMap.put(node, nodeElement)
+        elementsToPostProcess.add(new Pair(node, nodeElement))
+        nodeElement.trace(node)
```

Comment:
The direct tracing is completely removed from the generator on the master branch because of a performance issue, this should adapt to that as well for more accurate comparisons during evaluation

Reply:
Or you can for now turn off both tracing options text-diagram and diagram-text, as that also removes the performance overhead